File: Lab 4/pattern.py
def main():
    # Take user input for text and pattern
    text = input("Enter the text: ")
    pattern = input("Enter the pattern: ")

    textLen = len(text)
    patternLen = len(pattern)

    # Naive Pattern Matching Algorithm
    for i in range(textLen - patternLen + 1):
        j = 0

        # Check if pattern matches the substring starting at index i
        print(f"\nChecking substring starting at index {i}:")
        while j < patternLen and text[i + j] == pattern[j]:
            print(f"Comparing text[{i + j}] = '{text[i + j]}' with pattern[{j}] = '{pattern[j]}' => Match")
            j += 1

        # If pattern matches completely, print the index
        if j == patternLen:
            print(f"Pattern found at index {i}")
            break
        else:
            print(f"Pattern does not match starting at index {i}")

if __name__ == "__main__":
    main()

File: Lab 4/pattern.c
#include <stdio.h>
#include <string.h>

int main() {
    char text[1000], pattern[1000];

    printf("Enter the text: ");
    fgets(text, sizeof(text), stdin);
    text[strcspn(text, "\n")] = '\0';  

    printf("Enter the pattern: ");
    fgets(pattern, sizeof(pattern), stdin);
    pattern[strcspn(pattern, "\n")] = '\0';  

    int textLen = strlen(text), patternLen = strlen(pattern);

    for (int i = 0; i <= textLen - patternLen; i++) {
        printf("\nChecking substring at index %d: ", i);
        int j;
        for (j = 0; j < patternLen; j++) {
            printf("Comparing '%c' with '%c' -> %s\n", text[i + j], pattern[j], (text[i + j] == pattern[j]) ? "Match" : "No Match");
            if (text[i + j] != pattern[j]) break;
        }
        if (j == patternLen) {
            printf("Pattern found at index %d\n", i);
            return 0;  // Exit after first match
        }
    }

    printf("Pattern not found.\n");
    return 0;
}
